### PR TITLE
[AIRFLOW-679] Kill task if another is running concurrently

### PR DESCRIPTION
**Pull request by saguziel**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-679

Testing Done:
- Ran a task, set PID to be different, and ensuring it failed

If there's a connection error while heartbeating, it should retry. Also,
if it hasn't been able to heartbeat for a while, it should kill the
child processes so that we don't have 2 of the same task running.

@aoen @zodiac @plypaul

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -2025,22 +2027,24 @@ def on_kill(self):
...
+        new_ti = session.query(TI).filter(
             TI.dag_id==ti.dag_id, TI.task_id==ti.task_id,
             TI.execution_date==ti.execution_date).scalar()
-        if state == State.RUNNING:
+        if new_ti.state == State.RUNNING:
+            self.logger.info("running")
             self.was_running = True
+            if not (socket.getfqdn() == new_ti.hostname and self.process.pid == new_ti.pid):
+                logging.warning("Recorded hostname and pid of {new_ti.hostname} and {new_ti.pid}"
+                                "does not match this instance's which is {socket.getfqdn()}"
```
Comment (aoen) [inline airflow/jobs.py]: Nit: s/is/are

File: airflow/jobs.py [hunk 2]
```diff
@@ -2025,22 +2027,24 @@ def on_kill(self):
...
         # Suicide pill
         TI = models.TaskInstance
         ti = self.task_instance
-        state = session.query(TI.state).filter(
+        new_ti = session.query(TI).filter(
             TI.dag_id==ti.dag_id, TI.task_id==ti.task_id,
             TI.execution_date==ti.execution_date).scalar()
-        if state == State.RUNNING:
+        if new_ti.state == State.RUNNING:
+            self.logger.info("running")
```
Comment (aoen) [inline airflow/jobs.py]: This looks like a message you were using to debug

**Comment by aoen** [comment]
LGTM after nits and CI/landscape are fixed.

File: airflow/jobs.py [hunk 3]
```diff
@@ -2025,22 +2027,25 @@ def on_kill(self):
...
+                                "{self.process.pid}. Taking the poison pill. So "
+                                "long."
+                                .format(**locals()))
+                raise AirflowException("Another worker/process is running this job")
         elif self.was_running and hasattr(self, 'process'):
             logging.warning(
                 "State of this instance has been externally set to "
                 "{self.task_instance.state}. "
                 "Taking the poison pill. So long.".format(**locals()))
-            self.process.terminate()
```
Comment (aoen) [inline airflow/jobs.py]: Curious why the logic was changed from killing the PID to raising an exception?

Comment (saguziel) [reply airflow/jobs.py]: @plypaul added the logic in the finally block above to kill all child processes, I believe this does not do so.

File: airflow/jobs.py [hunk 4]
```diff
@@ -2032,11 +2038,19 @@ def heartbeat_callback(self, session=None):
...
-        state = session.query(TI.state).filter(
+        new_ti = session.query(TI).filter(
             TI.dag_id==ti.dag_id, TI.task_id==ti.task_id,
             TI.execution_date==ti.execution_date).scalar()
-        if state == State.RUNNING:
+        if new_ti.state == State.RUNNING:
             self.was_running = True
+            if not (socket.getfqdn() == new_ti.hostname and self.process.pid == new_ti.pid):
+                logging.warning("Recorded hostname and pid of {new_ti.hostname} "
+                                "and {new_ti.pid} does not match this instance's "
```
Comment (aoen) [inline airflow/jobs.py]: Nit: s/does/do

File: airflow/jobs.py [hunk 5]
```diff
@@ -2032,11 +2038,19 @@ def heartbeat_callback(self, session=None):
...
+                                "which are {socket.getfqdn()} and "
```
Comment (aoen) [inline airflow/jobs.py]: Nit: dedup socket.getfqdn() to a variable

**Comment by aoen** [comment]
Fix nits then LGTM.

File: airflow/jobs.py [hunk 6]
```diff
@@ -1983,6 +1983,13 @@ def _execute(self):
             )
             self.process = subprocess.Popen(['bash', '-c', command])
             self.logger.info("Subprocess PID is {}".format(self.process.pid))
+            ti = self.task_instance
```
Comment (bolkedebruin) [inline airflow/jobs.py:1986]: @saguziel @aoen you cannot depend on task_instance being fully populated. SQL Merging here made "try_number" always be 0, so retries will try forever. I have a PR out for that in a few minutes and hopefully a test with it. But I wonder what is happening in your production?

**Pull request closed without merging**